exchanges/orderbook/linked_list.go

```diff
@@ -0,0 +1,675 @@
+			if tip.value.ID == updts[x].ID {
+				if tip.value.Price != updts[x].Price { // Price level change
+					// bookmark tip to move this node to correct price level
+					bookmark = tip
+					continue // continue through node depth
+				}
+				// no price change, amend amount and conintue update
+				tip.value.Amount = updts[x].Amount
+				continue updates // continue to next update
+			}
+
+			if tip.value.Price < updts[x].Price {
+				if bookmark != nil { // shift bookmarked node to current tip
+					bookmark.value = updts[x]
+					move(&ll.head, bookmark, tip)
+					continue updates
+				}
+
+				// search for ID
+				for n := tip.next; n != nil; n = n.next {
+					if n.value.ID == updts[x].ID {
+						n.value = updts[x]
+						// inserting before the tip
+						move(&ll.head, n, tip)
+						continue updates
+					}
+				}
+				// ID not matched in depth so add correct level for insert
+				bookmark = tip
+				break
+			}
+
+			if tip.next == nil {
+				if bookmark == nil {
+					bookmark = tip
+				} else {
+					bookmark.value = updts[x]
+					bookmark.next.prev = bookmark.prev
+					if bookmark.prev == nil { // Bookmark is at head
+						ll.head = bookmark.next
+					} else {
+						bookmark.prev.next = bookmark.next
+					}
+					tip.next = bookmark
+					bookmark.prev = tip
+					bookmark.next = nil
+					continue updates
+				}
+			}
+		}
+		n := stack.Pop()
+		n.value = updts[x]
+		switch {
+		case bookmark.prev == nil:
+			n.next = ll.head
+			ll.head.prev = n
+			ll.head = n
+		case bookmark.next == nil:
+			bookmark.next = n
+			n.prev = bookmark
+		default:
+			bookmark.prev.next = n
+			n.prev = bookmark.prev
+			bookmark.prev = n
+			n.next = bookmark
+		}
+		ll.length++
+	}
+	return nil
+}
+
+// insertUpdates inserts new updates for bids based on price level
+func (ll *bids) insertUpdates(updts Items, stack *stack) error {
+	for x := range updts {
+		var prev *node
+		for tip := &ll.head; ; tip = &(*tip).next {
+			if *tip == nil { // Head
+				n := stack.Pop()
+				n.value = updts[x]
+				n.prev = prev
+				ll.length++
+				*tip = n
+				break // Continue updates
+			}
+
+			if (*tip).value.Price == updts[x].Price { // Price already found
+				return fmt.Errorf("%w for price %f",
+					errCollisionDetected,
+					updts[x].Price)
+			}
+
+			if (*tip).value.Price < updts[x].Price { // Alignment
+				n := stack.Pop()
+				n.value = updts[x]
+				n.prev = prev
+				ll.length++
+				// Reference current with new node
+				(*tip).prev = n
+				// Push tip to the right
+				n.next = *tip
+				// This is the same as prev.next = n
+				*tip = n
+				break // Continue updates
+			}
+
+			if (*tip).next == nil { // Tail
+				n := stack.Pop()
+				n.value = updts[x]
+				n.prev = prev
+				ll.length++
+				(*tip).next = n
+				n.prev = *tip
+				break // Continue updates
+			}
+			prev = *tip
+		}
+	}
+	return nil
+}
+
+// asks imbed a linked list to attach methods for ask depth specific
+// functionality
+type asks struct {
```

Comment:
Why is this like this instead of on the `Depth` struct ie
```
type Depth struct {
	asks linkedList
        bids linkedList
        ...
}
```

It looks as though you're trying to prevent a `bids` from ever calling an `asks` implementation while also re-using function names between them. I'm not saying to revert this, but I do find it strange.

There is a huge amount of reused code between the asks and bids functions and it would be good to extract that out into singular functions either way

Reply:
There are subtle differences in the functions and I wanted to make sure there was segregation between the ask and bid methods because of these differences. There is a lot of re-used code but we should go through everything with a fine tooth comb to figure out whats going to be inlined correctly. I prefer (which you can disagree) this expanded form for now to make sure everything is working as intended. Then we can look at breaking everything up. @thrasher-